### PR TITLE
Automatic Tuning API

### PR DESCRIPTION
**Pull request by tomapaunovic**
We want to add additional API for configuring SQL Azure DB advisors with single Auto Tuning API endpoint. Currently we have following endpoints
 
<primary-endpoint>/subscriptions/{SubscriptionId}/resourceGroups/{GroupName}/providers/Microsoft.Sql/servers/{ServerName}/advisors/{AdvisorName}

<primary-endpoint>/subscriptions/{SubscriptionId}/resourceGroups/{GroupName}/providers/Microsoft.Sql/servers/{ServerName}/databases/{databaseName}/advisors/{AdvisorName}
 
We want to add following endpoints
 
<primary-endpoint>/subscriptions/{SubscriptionId}/resourceGroups/{GroupName}/providers/Microsoft.Sql/servers/{ServerName}/automaticTuning/current

<primary-endpoint>/subscriptions/{SubscriptionId}/resourceGroups/{GroupName}/providers/Microsoft.Sql/servers/{ServerName}/databases/{databaseName}/automaticTuning
 
Full description of our API can be found here
 
https://microsoft.sharepoint.com/teams/Workload_Insight/Shared%20Documents/Automatic%20Tuning%20RP%20API.docx?web=1
 
New  API should exist on Database and Server level with following properties
 
• Auto tuning status, with following values 
o Inherit (on DB level only) – inherits advisors auto execute settings from parent server
o Auto – all advisors are automatically configured
o Custom – customer sets auto execute settings individually for each advisor

• Per advisors status for each advisors with On/Off/Default values. On/Off means specific per-advisor override while Default means advisor settings as specfied by Auto/Inherit automatic tuning status. If automatic tuning status is set to Custom then Default is not valid per advisor option
 
In T-SQL terms, we want to replicate following DDL (that will be added as part of same effort)
 
ALTER DATABASE My DB SET AUTOMATIC_TUNING = INHERIT | AUTO |CUSTOM

ALTER DATABASE My DB SET AUTOMATIC_TUNING ( 
     CREATE_INDEX = ON,  
     DROP_INDEX = OFF, 
     PLAN_CORRECTION = DEFAULT
     )

---
This change is [Reviewable](https://reviewable.io/reviews/azure/azure-rest-api-specs/1198)

### CODE REVIEW

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: json is not valid - missing "comma" https://travis-ci.org/Azure/azure-rest-api-specs/jobs/229880407#L2188

**Review by veronicagg — CHANGES_REQUESTED** [review]

**Comment by veronicagg** [comment]
Please provide x-ms-examples for the operations, see https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist for validation and documentation purposes.

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: should there be a "schema" for the default case?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: is this operation under the proper path?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: is this operation under the appropriate path?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: malformed swagger, which definition should this be in?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Please check model definition name.

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: this model is not defined in the spec

**Review by veronicagg — CHANGES_REQUESTED** [review]

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Arm.Tools project, that is used for generating swagger in Arm-SQL, only populates error description under default

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Removed

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Removed

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Removed

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Removed

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Removed, it's a leftover from server operations

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Removed as leftover from VirtualNetwork API

**Comment by veronicagg** [comment]
@tomapaunovic thanks for the replies, I'll review the changes once your new commit is pushed.

**Comment by veronicagg** [comment]
@tomapaunovic I haven't seen a new commit, when are you planning on sending it?

**Comment by veronicagg** [comment]
@tomapaunovic there's a syntax error in your spec, please take a look at https://travis-ci.org/Azure/azure-rest-api-specs/jobs/233257266#L2420

**Review by veronicagg — CHANGES_REQUESTED** [review]
please see my comment.

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: it's ok to have a description here, but I'm wondering what models should be used for the errors, see https://github.com/Azure/autorest/blob/bf30da3ae09718d336a1eca49afd7d053d83657e/docs/developer/guide/defining-clients-swagger.md#error-modeling

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: see my reply to a previous comment about error modeling https://github.com/Azure/autorest/blob/bf30da3ae09718d336a1eca49afd7d053d83657e/docs/developer/guide/defining-clients-swagger.md#error-modeling

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: will you be adding an x-ms-example for this one?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: will you be adding an x-ms-example for this one?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: minor: fix indenting

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: should not remove name of the model?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: please use x-ms-enum see https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-enum

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: this model is defined as a list, is that appropriate? the model is defined as a pageable model

**Review by veronicagg — CHANGES_REQUESTED** [review]

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (nathannfan) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: @veronicagg is this a new requirement?  We haven't been asked for this before.

Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: I'm not sure if we have support to generate error models in SQL Web API. I've started a thread with Nathan to discuss this issue

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Server level automatic tuning is still under development, so I will add x-ms-examples when I have a working prototype. Database level automatic tuning is fully functional, and those examples are generated from real request/response

Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: good to know, only APIs deployed to ARM should be included in a merged spec, if you'd like to have the spec approved and merged for the APIs that are already deployed, you can split this one into another PR. If you expect this API will be deployed soon, we can hold the PR a bit, until it's deployed, let me know. thanks!

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: missing "modelAsString": true/false? see https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-enum

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: missing "modelAsString" property

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: missing "modelAsString" property

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: missing "modelAsString"

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: is this a proper description?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: should location be required?

**Review by veronicagg — CHANGES_REQUESTED** [review]
Thanks for the updates @tomapaunovic . 
Few more comments inline. 
Also, should you be adding security definition: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/233921730#L1146 (see https://github.com/Azure/autorest/blob/master/docs/developer/validation-rules/security-definitions-structure-validation.md for more info).
Are there operations in the spec that are not deployed to ARM? I see that we're still missing the x-ms-examples for the operations referred to in the earlier comments. Please let me know, as we should not merge the spec with endpoints that aren't working. Once the spec if merged, SDKs can be generated out of it, and they are expected to work.

**Comment by tomapaunovic** [comment]
None of these changes are deployed to ARM. If I understand this process correctly, swagger file needs to be commited to azure-rest-api-specs depot before I check-in. dabase/automaticTuning/current APIs are working fully and I'm this PR is the only thing that is blocking me to check-in those changes. server/automaticTuning/current are still work in progress, but I wanted to do swagger part first since APIs are really similar. So after server/automaticTuning/current APIs are working I can provide examples. Does that works fine for you?

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: ModelAsString is set in code to false, but it's not visible in swagger. It looks like swagger generator generates modelAsString only in cases when it's set to true. I will start a thread with Nathan to discuss this issue

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: All resources generated in SQL Web API have same properties description. To view actual description you have to go to referenced type segment, in this case to AutomaticTuningProperties

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (tomapaunovic) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Actually TrackedResource is not used by this API, so I will remove it

**Comment by tomapaunovic** [comment]
SecurityDefinitions are currently not generated from SQL Web API. I will start a separate thread with Nathan to discuss this issue

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (nathannfan) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: This is another issue that has never been brought up with us before.  I'll modify the generator since now we want explicit "modelAsString": "false"

Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: @tomapaunovic ok, if you want it set to false, it should be fine then. Just to confirm, you would like this property to be modeled as an Enum type in C# for example (for awareness, in this case if you were to add another value later on, you would be making a breaking change).

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Ok, just wanted to double check to make sure if was a conscious decision on your side.

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Ok, thanks!

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (nathannfan) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: Got it.  We have a required property in our generator attribute where it adds modelAsString:true if true, and nothing if false.  I did it that way because it seemed to be the precedent, but I see how it might seem like we aren't fully aware of the property.

I checked in a change to our Swashbuckle project to do explicit modelAsString:false to make this more clear in our docs.  Thanks, Veronica.

@tomapaunovic you probably already did this, but if you haven't, please make sure all the enums not marked modelAsString are not likely to change.

**Comment by veronicagg** [comment]
@tomapaunovic regarding the process, I saw clarification offline on this, so I hope we're good on that. On my side, we should not merge the spec unless the APIs included are deployed. 
Regarding security definition, if you can add it great, if not we could postpone that one for a later PR. 
You mentioned above you'll be updating the spec to remove the TrackedResource section that's not used, so I'll wait for that commit. If the APIs are deployed at the time of that commit, we could merge the spec, if not, we could just agree on the spec being good and hold the merge till you deploy them.

**Comment by jaredmoo** [comment]
Hi Veronica, do you mind if we keep TrackedResource there? Here's the reason - my understanding is that it's better to consolidate into a single swagger file so that we aren't using composite swagger, and when we do this all of the common definitions (such as TrackedResource) will obviously be used in that consolidated file. So I don't want to waste effort updating our generator logic to remove these common definitions if they happen to be unused in that file. Sound good?

**Review by jaredmoo — APPROVED** [review]
Approved but can't be merged until this is available in production.

File: arm-sql/2015-05-01-preview/swagger/automaticTuning.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/swagger/automaticTuning.json]: minor: is indenting proper here?

File: arm-sql/2015-05-01-preview/examples/DatabaseAutomaticTuningGet.json
Comment (veronicagg) [inline arm-sql/2015-05-01-preview/examples/DatabaseAutomaticTuningGet.json]: out of curiosity, is this service considered as preview on your end, so you're following the convention of using "-preview" on the api version?

**Review by veronicagg — CHANGES_REQUESTED** [review]
left couple of minor question/comments. 
Rest looks good, won't approve until we're ready to merge, to avoid confusion, so let us know when your service it deployed and we can approve and merge, thanks!

**Comment by salameer** [comment]
Captain @jaredmoo how are we doing here :)?

If we have no more changes requested can we mark it as approved for now?

**Comment by jaredmoo** [comment]
@tomapaunovic do you know when this API will be in production?

**Comment by jaredmoo** [comment]
Let me check when it will be in production.

**Comment by veronicagg** [comment]
@jaredmoo @tomapaunovic is the service now deployed?

**Comment by veronicagg** [comment]
@jaredmoo @tomapaunovic any updates? If we don't her back on this one in the next day or so, we would close this PR and you can reopen when you're back to it. thanks!

**Comment by jaredmoo** [comment]
This will not be in production for a few more weeks still. Please close for now if you wish, we will comment when to reopen. @tomapaunovic is just extremely efficient and sent this PR well in advance :)

**Comment by veronicagg** [comment]
@jaredmoo sounds good! it's good that he got feedback and fixes in ahead of time, just reopen when you're ready, thanks!

**Pull request closed without merging**
